---
title: "R Notebook"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. 

```{r}
plot(cars)
```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Cmd+Option+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Cmd+Shift+K* to preview the HTML file). 

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.

# Import the library
```{r}
library(scPipe)
library(SingleCellExperiment)
data_dir = "sra-toolkit/"
```

# Load predefined files
```{r}
# file path:
hg38fa_fn = file.path(data_dir, "hg38.fa.gz")
hg38anno_fn = file.path(data_dir, "hg38_anno.gff3.gz")
barcode_annotation_fn = system.file("extdata", "barcode_anno.csv", package = "scPipe")
```

```{r}
sc_trim_barcode(file.path(data_dir, "combined.fastq.gz"),
                file.path(data_dir, "SRR7706271_1.fastq"),
                file.path(data_dir, "SRR7706271_2.fastq"),
                read_structure = list(bs1=-1, bl1=0, bs2=6, bl2=8, us=0, ul=6))
```


# 1. Transform fastq file into bam file
```{r}
if(.Platform$OS.type != "windows"){
  Rsubread::buildindex(basename=file.path(data_dir, "hg38_index"), reference=hg38fa_fn)

  Rsubread::align(index=file.path(data_dir, "hg38_index"),
      readfile1=file.path(data_dir, "combined.fastq.gz"),
      output_file=file.path(data_dir, "out.aln.bam"), phredOffset=64)
}
```


#2. Assigning reads to annotated exons
```{r}
if(.Platform$OS.type != "windows"){
  sc_exon_mapping(file.path("sra-toolkit/out.aln.bam"),
                file.path("sra-toolkit/out.map.bam"),
                ERCCanno_fn)
}
```

#3. De-multiplexing data and counting genes
```{r}
if(.Platform$OS.type != "windows"){
  sc_demultiplex(file.path("sra-toolkit/out.map.bam"), data_dir, barcode_annotation_fn,has_UMI=TRUE)
  sc_gene_counting("sra-toolkit/", barcode_annotation_fn)
}
```

