---
title: "R Notebook"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. 

```{r}
plot(cars)
```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Cmd+Option+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Cmd+Shift+K* to preview the HTML file). 

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.

# Import the library
```{r}
library(scPipe)
library(SingleCellExperiment)
data_dir = tempdir()
```

# Load predefined files
```{r}
# file path:
ERCCfa_fn = system.file("extdata", "ERCC92.fa", package = "scPipe")
ERCCanno_fn = system.file("extdata", "ERCC92_anno.gff3", package = "scPipe")
barcode_annotation_fn = system.file("extdata", "barcode_anno.csv", package = "scPipe")

fq_R1 = system.file("extdata", "simu_R1.fastq.gz", package = "scPipe")
fq_R2 = system.file("extdata", "simu_R2.fastq.gz", package = "scPipe")
```

# 1. Transform fastq file into bam file
```{r}
if(.Platform$OS.type != "windows"){
  Rsubread::buildindex(basename=file.path(data_dir, "ERCC_index"), reference=ERCCfa_fn)
  
  Rsubread::align(index=file.path(data_dir, "ERCC_index"),
                  readfile1=file.path("sra-toolkit/SRR7547688.fastq"),
                  output_file=file.path("sra-toolkit/out.aln.bam"), phredOffset=64)
}
```


#2. Assigning reads to annotated exons
```{r}
if(.Platform$OS.type != "windows"){
  sc_exon_mapping(file.path("sra-toolkit/out.aln.bam"),
                file.path("sra-toolkit/out.map.bam"),
                ERCCanno_fn)
}
```

#3. De-multiplexing data and counting genes
```{r}
if(.Platform$OS.type != "windows"){
  sc_demultiplex(file.path("sra-toolkit/out.map.bam"), data_dir, barcode_annotation_fn,has_UMI=TRUE)
  sc_gene_counting("sra-toolkit/", barcode_annotation_fn)
}
```


